### PR TITLE
Add a new preferences http handler for internal usage

### PR DESCRIPTION
**Pull request by wyzhang**
Current handler returns Map<String, String>. Creating a new one for internal usage which returns PreferencesDetail that encapsulates the Map along with some other info.

This is server side of change in preparation of switching ProfileMetadataMessageProcessor from direct levelDB access to via REST API calls.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java [hunk 1]
```diff
@@ -64,15 +65,27 @@ public PreferencesService(MessagingService messagingService,
...
-  private Map<String, String> getConfigProperties(EntityId entityId) {
+  private PreferencesDetail get(EntityId entityId) {
     return TransactionRunners.run(transactionRunner, context -> {
       return new PreferencesTable(context).getPreferences(entityId);
     });
   }
 
+  private PreferencesDetail getResolved(EntityId entityId) {
+    return TransactionRunners.run(transactionRunner, context -> {
+        return new PreferencesTable(context).getResolvedPreferences(entityId);
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: style: indentation

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java [hunk 2]
```diff
@@ -64,15 +65,27 @@ public PreferencesService(MessagingService messagingService,
...
+    });
+  }
+
+  private Map<String, String> getConfigProperties(EntityId entityId) {
+    return TransactionRunners.run(transactionRunner, context -> {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: can return `get(entityId).getProperties()`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java [hunk 3]
```diff
@@ -166,58 +179,66 @@ private void deleteConfig(EntityId entityId) {
...
+  public PreferencesDetail getPreferences() { return get(new InstanceId("")); }
+
 
   /**
    * Get namespace level preferences
    */
-  public Map<String, String> getProperties(NamespaceId namespaceId) {
-    return getConfigProperties(namespaceId);
-  }
+  public Map<String, String> getProperties(NamespaceId namespaceId) { return getConfigProperties(namespaceId); }
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: it doesn't seem like we should have these methods anymore, with callers just getting them from the PreferencesDetail.

If this would affect a whole bunch of classes, we can also do this in a separate PR.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java [hunk 4]
```diff
@@ -166,58 +179,66 @@ private void deleteConfig(EntityId entityId) {
...
+
+  public PreferencesDetail getPreferences(NamespaceId namespaceId) { return get(namespaceId); }
+
 
   /**
    * Get app level preferences
    */
   public Map<String, String> getProperties(ApplicationId applicationId) {
     return getConfigProperties(applicationId);
   }
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java:205]: style: newline between methods

File: cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  /**
+   * Sequence id of operations on the preferences
+   */
+  private final Long seqId;
+  /**
+   * Whether it is a resolved preferences or not.
+   */
+  private boolean resolved;
+
+  public static PreferencesDetail merge(PreferencesDetail left, PreferencesDetail right) {
```
Comment (albertshau) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java]: can you add a javadoc to this method, it's not immediately obvious what merge means

File: cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    if (left == null && right == null) {
```
Comment (albertshau) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java]: by convention, anything that isn't annotated as `@Nullable` is assumed to be non-null, and in general we try to avoid null values when possible. When would be trying to merge a null?

File: cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+      seqId = new Long(leftSeqId.longValue());
+    } else if (rightSeqId != null) {
+      seqId = new Long(rightSeqId.longValue());
+    }
+
+    boolean resolved = (left.resolved || right.resolved);
+    return new PreferencesDetail(properties, seqId, resolved);
+  }
+
+  public PreferencesDetail(Map<String, String> properties, @Nullable Long seqId, boolean resolved) {
```
Comment (albertshau) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java]: I don't think we should allow seqId to be null, as that isn't a valid representation of stored preferences. It forces the caller of getSeqId() to handle a null, which shouldn't ever be the case

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/PreferencesHttpHandlerInternal.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    this.preferencesService = preferencesService;
+    this.applicationLifecycleService = applicationLifecycleService;
+    this.namespaceQueryAdmin = namespaceQueryAdmin;
+  }
+
+  @Path("/preferences")
+  @GET
+  public void getInstancePreferences(HttpRequest request, HttpResponder responder) {
+    PreferencesDetail detail = preferencesService.getPreferences();
+    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(detail, PreferencesDetail.class));
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/PreferencesHttpHandlerInternal.java:71]: I believe sendJson will serialize the object to json automatically, which mean you don't need the GSON.toJson()

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/PreferencesHttpHandlerInternal.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+  }
+
+  @Path("/namespaces/{namespace-id}/apps/{application-id}/{program-type}/{program-id}/preferences")
+  @GET
+  public void getProgramPreferences(HttpRequest request, HttpResponder responder,
+                                    @PathParam("namespace-id") String namespace, @PathParam("application-id") String appId,
+                                    @PathParam("program-type") String programType, @PathParam("program-id") String programId,
+                                    @QueryParam("resolved") boolean resolved) throws Exception {
+    ProgramId program = new ProgramId(namespace, appId, getProgramType(programType), programId);
+    applicationLifecycleService.getAppDetail(program.getParent());
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/PreferencesHttpHandlerInternal.java]: this should be returning a 404 if the app exists but the program does not

File: cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java (hunk 5 above)
Comment (albertshau) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java]: I think it's more descriptive to name the method resolve(), with 'parent' and 'child' instead of 'left' and 'right'.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java (hunk 3 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: Correct. Added TODO. Clean up in separate PR

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java (hunk 2 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java (hunk 1 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java (hunk 4 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java:205]: Fixed.

File: cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java (hunk 5 above)
Comment (wyzhang) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java]: Agreed. Done.

File: cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java (hunk 6 above)
Comment (wyzhang) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java]: We don't such case of resolve(null, non-null), but added code to protect against that. 
Removed.

File: cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java (hunk 7 above)
Comment (wyzhang) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/PreferencesHttpHandlerInternal.java (hunk 8 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/PreferencesHttpHandlerInternal.java:71]: void sendJson(HttpResponseStatus status, String jsonString);

Don't we still need GSON.toJson() to convert to jsonString?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/PreferencesHttpHandlerInternal.java (hunk 9 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/PreferencesHttpHandlerInternal.java]: this can be done with Store.ensureProgramExists()

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/PreferencesHttpHandlerInternal.java (hunk 8 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/PreferencesHttpHandlerInternal.java:71]: oh this is the netty responder, got this mixed up with HttpServiceResponder. Ignore my comment.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/PreferencesHttpHandlerInternal.java (hunk 9 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/PreferencesHttpHandlerInternal.java]: Done

File: cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java (hunk 7 above)
Comment (wyzhang) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java]: Actually seqId could be null when preferences never get set on the entity. Given the fact that seqId > 0, I am using 0 as default value when preferences are not present. This is consistent with current contact that property map is empty when preferences are not set.

Comment (albertshau) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java]: I see, makes sense

File: cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java [hunk 10]
```diff
@@ -28,55 +27,47 @@
...
-    }
+    Long seqId = getMaxSeqId(parent.getSeqId(), child.getSeqId());
 
-    boolean resolved = (left.resolved || right.resolved);
-    return new PreferencesDetail(properties, seqId, resolved);
+    return new PreferencesDetail(properties, seqId, true);
   }
 
-  public PreferencesDetail(Map<String, String> properties, @Nullable Long seqId, boolean resolved) {
+  public PreferencesDetail(Map<String, String> properties, Long seqId, boolean resolved) {
```
Comment (albertshau) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java]: if it can't be null, it should be a long instead of a Long

File: cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java [hunk 11]
```diff
@@ -105,5 +96,9 @@ public String toString() {
       "resolved='" + resolved +
       '}';
   }
+
+  public static Long getMaxSeqId(Long left, Long right) {
```
Comment (albertshau) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java]: there isn't much value in this method, can just call Long.max directly

File: cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java (hunk 7 above)
Comment (albertshau) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java]: I see. Does this cause any issues if preferences are set and then deleted? If the sequence number gets reset in that scenario, it would mean some of the assumptions in that metadata processor are false.

Comment (wyzhang) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java]: Deletion just clear the map but still bump up the seqId. So the only case seqId is not there is the entity never had any preferences set on it.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/PreferencesHttpHandlerInternal.java (hunk 8 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/PreferencesHttpHandlerInternal.java:71]: Resolved.

File: cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java (hunk 11 above)
Comment (wyzhang) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java]: Done.

File: cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java (hunk 10 above)
Comment (wyzhang) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/PreferencesDetail.java]: Done.

**Review by albertshau — APPROVED** [review]

**Review by albertshau — APPROVED** [review]

**Comment by wyzhang** [comment]
https://builds.cask.co/browse/CDAP-DUT7115-5

Build and test passed.
